azure-iot-device/azure/iot/device/provisioning/sk_registration_client.py

```diff
@@ -0,0 +1,80 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+"""
+This module contains one of the implementations of the Registration Client which uses Symmetric Key authentication.
+"""
+import logging
+from threading import Event
+from .registration_client import RegistrationClient
+from .models.registration_result import RegistrationResult
+from .internal.polling_machine import PollingMachine
+
+logger = logging.getLogger(__name__)
+
+
+class SymmetricKeyRegistrationClient(RegistrationClient):
+    """
+    Client which can be used to run the registration of a device using Symmetric Key authentication.
+    """
+
+    def __init__(self, mqtt_state_based_provider):
+        """
+        Initializer for the Symmetric Key Registration Client
+        """
+        super(SymmetricKeyRegistrationClient, self).__init__(mqtt_state_based_provider)
+        self._polling_machine = PollingMachine(mqtt_state_based_provider)
+        # To be defined by sample
+        self.on_registration_update = None
+        self._polling_machine.on_connected = self._on_state_change
+        self._polling_machine.on_disconnected = self._on_state_change
+        self._polling_machine.on_registration_status_update = self._on_device_registration_update
+
+    def register(self):
+        """
+        Register the device with the provisioning service.
+        """
+        logger.info("Registering with Hub...")
+        register_complete = Event()
+
+        def callback_register(result):
+            register_complete.set()
```

Comment:
>           register_complete.set() [](start = 1, length = 34)

nit: this should be at the end of callback_register() instead of the beginning.  I don't think it will cause any problems here, but you're doing "notify that caller that the operation is complete" followed by "do a few more things" and I've seen that cause hard-to-reproduce bugs before depending on what "a few more things" are.  It's better to have the notification as the very last step.